File: src/core/main/VGraphics.java
package core.main;

import core.main.structs.Vector;
import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Composite;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics2D;
import java.awt.Shape;
import java.awt.Stroke;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class VGraphics {

    private Graphics2D g;
    private ArrayList<Graphics2D> graphicsStack;
    private ArrayList<AffineTransform> saves;
    
    public VGraphics(Graphics2D g){
        graphicsStack = new ArrayList<>();
        graphicsStack.add(g);
        this.g = g;
        saves = new ArrayList<>();
    }
    
    public void save(){ saves.add(g.getTransform()); }
    public void reset(){ g.setTransform(saves.remove(saves.size()-1)); }
    
    public void setTransform(AffineTransform at){ g.setTransform(at); }
    public AffineTransform getTransform(){ return g.getTransform(); }
    
    public void subGraphics(Graphics2D g){ graphicsStack.add(g); this.g = g; }
    public void resetGraphics(){ 
        graphicsStack.remove(graphicsStack.size()-1);
        this.g = graphicsStack.get(graphicsStack.size()-1); 
    }
    
    public FontMetrics getFontMetrics(){ return g.getFontMetrics(); }
    public Composite getComposite(){ return g.getComposite(); }
    
    public void translate(Vector v){ g.translate(v.x, v.y); }
    public void scale(double d){ g.scale(d, d); }
    public void scale(Vector v){ g.scale(v.x, v.y); }
    public void rotate(double a){ g.rotate(a); }
    public void rotate(double a, Vector p){ g.rotate(a, p.x, p.y); }
    public void transform(AffineTransform at){ g.transform(at); }
    
    public void setComposite(Composite c){ g.setComposite(c); }
    public void setStroke(Stroke s){ g.setStroke(s); }
    public void resetStroke(){ g.setStroke(new BasicStroke(1)); }
    public void setColor(Color c){ g.setColor(c); }
    public void setFont(Font f){ g.setFont(f); }
    
    public void draw(Shape s){ g.draw(s); }
    public void fill(Shape s){ g.fill(s); }
    
    public void drawImage(BufferedImage img, Vector pos){ g.drawImage(img, (int)pos.x, (int)pos.y, null); }
    public void drawImage(BufferedImage img, Vector pos, Vector size){ g.drawImage(img, (int)pos.x, (int)pos.y, (int)size.x, (int)size.y, null); }
    public void drawImage(BufferedImage img, AffineTransform at){ g.drawImage(img, at, null); }
    
    public void drawString(String s, Vector pos){ g.drawString(s, (float)pos.x, (float)pos.y); }
}

File: src/velvet/main/VGraphics.kt
package velvet.main

import velvet.game.graphics.Sprite
import velvet.util.types.VColor
import velvet.util.types.spatial.Vector
import velvet.util.types.VShape
import java.awt.*
import java.awt.geom.AffineTransform
import java.awt.image.BufferedImage
import java.util.*

class VGraphics(private var g: Graphics2D) {

    private val graphicsStack: ArrayList<Graphics2D> = ArrayList()
    private val saves: ArrayList<AffineTransform>

    val baseGraphics: Graphics2D
        get() = g

    var transform: AffineTransform
        get() = g.transform
        set(at) { g.transform = at }

    val fontMetrics: FontMetrics
        get() = g.fontMetrics

    var font: Font
        get() = g.font
        set(value){ g.font = value }

    var composite: Composite
        get() = g.composite
        set(c) { g.composite = c }

    var stroke: Stroke
        get() = g.stroke
        set(value){ g.stroke = value }

    var color: VColor
        get() = VColor.fromJavaColor(g.color)
        set(value){ g.color = value.javaColor }

    init {
        graphicsStack.add(g)
        saves = ArrayList()
    }

    fun save() = saves.add(g.transform)
    fun reset(){ g.transform = saves.removeAt(saves.size - 1) }
    fun getTransformSave(index: Int) = saves[index]

    fun subGraphics(g: Graphics2D) {
        graphicsStack.add(g)
        this.g = g
    }

    fun resetGraphics() {
        graphicsStack.removeAt(graphicsStack.size - 1)
        this.g = graphicsStack.last()
    }

    fun translate(v: Vector) = g.translate(v.x, v.y)
    fun scale(d: Double) = g.scale(d, d)
    fun scale(v: Vector) = g.scale(v.x, v.y)
    fun rotate(a: Double) = g.rotate(a)
    fun rotate(a: Double, p: Vector) = g.rotate(a, p.x, p.y)
    fun transform(at: AffineTransform) = g.transform(at)

    fun resetStroke() { stroke = BasicStroke(1f) }
    fun setRenderingHint(key: RenderingHints.Key, value: Any) = g.setRenderingHint(key, value)

    fun draw(s: VShape) = g.draw(s.javaShape)
    fun fill(s: VShape) = g.fill(s.javaShape)

    fun drawImage(img: BufferedImage) = g.drawImage(img, 0, 0, null)
    fun drawImage(img: BufferedImage, pos: Vector) = g.drawImage(img, pos.x.toInt(), pos.y.toInt(), null)
    fun drawImage(img: BufferedImage, at: AffineTransform) {
        save()
        transform(at)
        drawImage(img)
        reset()
    }

    fun drawSprite(sprite: Sprite) = drawImage(sprite.image)
    fun drawSprite(sprite: Sprite, pos: Vector) = drawImage(sprite.image, pos)
    fun drawSprite(sprite: Sprite, at: AffineTransform) = drawImage(sprite.image, at)

    fun drawString(s: String, pos: Vector) = g.drawString(s, pos.x.toFloat(), pos.y.toFloat())
}